client/src/main/java/com/netflix/conductor/client/automator/TaskPollExecutor.java

```diff
@@ -0,0 +1,246 @@
+import com.netflix.appinfo.InstanceInfo.InstanceStatus;
+import com.netflix.conductor.client.http.TaskClient;
+import com.netflix.conductor.client.telemetry.MetricsContainer;
+import com.netflix.conductor.client.worker.PropertyFactory;
+import com.netflix.conductor.client.worker.Worker;
+import com.netflix.conductor.common.metadata.tasks.Task;
+import com.netflix.conductor.common.metadata.tasks.TaskResult;
+import com.netflix.conductor.common.utils.RetryUtil;
+import com.netflix.discovery.EurekaClient;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.concurrent.BasicThreadFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Manages the threadpool used by the workers for execution and server communication (polling and task update).
+ */
+class TaskPollExecutor {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(TaskPollExecutor.class);
+
+    private EurekaClient eurekaClient;
+    private TaskClient taskClient;
+    private int updateRetryCount;
+    private ExecutorService executorService;
+    private PollingSemaphore pollingSemaphore;
+
+    private static final String DOMAIN = "domain";
+    private static final String ALL_WORKERS = "all";
+
+    TaskPollExecutor(EurekaClient eurekaClient, TaskClient taskClient, int threadCount, int updateRetryCount,
+        String workerNamePrefix) {
+        this.eurekaClient = eurekaClient;
+        this.taskClient = taskClient;
+        this.updateRetryCount = updateRetryCount;
+
+        LOGGER.info("Initialized the TaskPollExecutor with {} threads", threadCount);
+
+        AtomicInteger count = new AtomicInteger(0);
+
+        this.executorService = Executors.newFixedThreadPool(threadCount,
+            new BasicThreadFactory.Builder()
+                .namingPattern(workerNamePrefix + count.getAndIncrement())
+                .uncaughtExceptionHandler(uncaughtExceptionHandler)
+                .build());
+
+        this.pollingSemaphore = new PollingSemaphore(threadCount);
+    }
+
+    void pollAndExecute(Worker worker) {
+        if (eurekaClient != null && !eurekaClient.getInstanceRemoteStatus().equals(InstanceStatus.UP)) {
+            LOGGER.debug("Instance is NOT UP in discovery - will not poll");
+            return;
+        }
+
+        if (worker.paused()) {
+            MetricsContainer.incrementTaskPausedCount(worker.getTaskDefName());
+            LOGGER.debug("Worker {} has been paused. Not polling anymore!", worker.getClass());
+            return;
+        }
+
+        Task task;
+        try {
+            if (!pollingSemaphore.canPoll()) {
+                return;
```

Comment:
A metric here to see how often we cannot poll?

Reply:
Same as above, since when workers are processing tasks, the poller will not be active so metric would not give any valuable info here.